proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientConfigSessionHandler.java

```diff
@@ -117,7 +118,8 @@ public boolean handle(PluginMessage packet) {
         server.getEventManager().fireAndForget(new PlayerClientBrandEvent(player, brand));
         player.setClientBrand(brand);
         // Client sends `minecraft:brand` packet immediately after Login,
-        // but at this time the backend server may not be ready, just discard it.
+        // but at this time the backend server may not be ready
+        brandPacket = packet.retain();
```

Comment:
Actually, if we retain, we should probably release this inside of the handlers inactive method

Reply:
Which method are you referring to exactly?

Reply:
I think that the better solution would be not to try to retain this, but to just send a new one, otherwise, would need to be inside of deactivate and handleDIsconnect

Reply:
Should I recreate the packet entirely or can I just .copy() it?

Reply:
```java
  public CompletableFuture<Void> handleBackendFinishUpdate(VelocityServerConnection serverConn) {
    String brand = serverConn.getPlayer().getClientBrand();
    if (brand != null) {
      PluginMessage brandPacket = new PluginMessage("minecraft:brand", Unpooled.wrappedBuffer(brand.getBytes()));

      serverConn.ensureConnected().write(
              PluginMessageUtil.rewriteMinecraftBrand(brandPacket.retain(), server.getVersion(),
                      player.getProtocolVersion()));
    }

    player.getConnection().write(new FinishedUpdate());
    serverConn.ensureConnected().write(new FinishedUpdate());
    return configSwitchFuture;
  }
  ```
  How about this?

Reply:
Yes, that generally looks better, we don't need to worry about ensuring that we don't leak the brand packet we got

Reply:
Pushed changes, I'm not sure if I need to use .retain() or not here though, and if it's ok to store the player's channel brand in that class.

Reply:
I don't think that calling retain is needed, but, I'm not 100% sure; storing the chanel name should be fine, we should only really see the minecraft:brand one in there, but ¯\_(ツ)_/¯

Reply:
I've removed it. Doesn't seem to be necessary after testing.